index.js

```diff
@@ -131,6 +131,36 @@ class ScmBase {
     stats() {
         return {};
     }
+
+    /**
+     * Return a unique identifier for the scmUrl
+     * @method getRepoId
+     * @param  {Object}    config        Configuration
+     * @param  {String}    config.scmUrl The scmUrl to generate ID
+     * @param  {String}    config.token  The token used to authenticate to the SCM
+     * @return {Promise}
+     */
+    getRepoId(config) {
+        const result = Joi.validate(config, dataSchema.plugins.scm.getRepoId);
+
+        if (result.error) {
+            return Promise.reject(result.error);
```

Comment:
I'm seeing a pattern a lot in this code, e.g. Check validation and reject on failure. Seems like a really good thing to refactor so these function's promise chains are easier to understand.

e.g.
```
validate(obj, schema) {
    const result = Joi.validate(obj, schema);
     
    if (result.error) {
         return Promise.reject(result.error);
    }
    
    return Promise.resolve(obj);
}
```

So you'd end up with something much less complex:
```
return validate(config, dataSchema.plugins.scm.getRepoId)
    .then(config => this._getRepoId(config))
    .then(repo => validate(repo, dataSchema.core.scm.repo);
```